File: examples/skill_md_creator/gemini_codegen/converter.js
const fs = require('fs');

class Converter {
    static convertCurlyPromptToSKILL(source_prompt_file, target_md_file) {
        const input = fs.readFileSync(source_prompt_file, 'utf8');
        const lines = input.split(/\r?\n/);

        let inSkill = false;
        let skillLines = [];
        let depth = 0;

        for (let line of lines) {
            const trimmed = line.trim();
            if (!inSkill) {
                if (trimmed.match(/^skill\s*\{/)) {
                    inSkill = true;
                    depth = 1;
                }
                continue;
            }

            if (trimmed === '{') depth++;
            else if (trimmed === '}') depth--;

            if (depth === 0) break;
            if (depth >= 1) skillLines.push(line);
        }

        if (skillLines.length === 0) return;

        let frontmatter = {};
        let contentLines = [];
        let i = 0;

        while (i < skillLines.length) {
            let line = skillLines[i].trim();
            if (!line) { i++; continue; }

            if (line.match(/^(header|head)\s*\{/)) {
                i = this.parseFrontmatterBlock(skillLines, i + 1, frontmatter);
            } else if (line.match(/^(content|body)\s*\{/)) {
                contentLines = this.extractBlock(skillLines, i);
                break;
            } else if (line.includes(':') && !line.match(/^(header|head|content|body|meta)\s*\{/)) {
                this.parseKV(line, frontmatter);
                i++;
            } else if (line.match(/^meta\s*\{/)) {
                i = this.parseMetaBlock(skillLines, i + 1, frontmatter);
            } else {
                i++;
            }
        }

        let output = '---\n';
        for (const [key, val] of Object.entries(frontmatter)) {
            if (typeof val === 'object') {
                output += `${key}:\n`;
                for (const [mk, mv] of Object.entries(val)) {
                    output += `  ${mk}: ${mv}\n`;
                }
            } else {
                output += `${key}: ${val}\n`;
            }
        }
        output += '---\n\n';

        output += this.processContent(contentLines);
        fs.writeFileSync(target_md_file, output.trimEnd() + '\n');
    }

    static parseKV(line, obj) {
        const idx = line.indexOf(':');
        if (idx === -1) return;
        const key = line.substring(0, idx).trim();
        const val = line.substring(idx + 1).trim().replace(/"/g, '');
        obj[key] = val;
    }

    static parseFrontmatterBlock(lines, start, obj) {
        let i = start;
        let d = 1;
        while (i < lines.length && d > 0) {
            let l = lines[i].trim();
            if (l === '{') d++;
            else if (l === '}') d--;
            else if (d === 1) {
                if (l.match(/^meta\s*\{/)) i = this.parseMetaBlock(lines, i + 1, obj) - 1;
                else this.parseKV(l, obj);
            }
            i++;
        }
        return i;
    }

    static parseMetaBlock(lines, start, obj) {
        if (!obj.meta) obj.meta = {};
        let i = start;
        let d = 1;
        while (i < lines.length && d > 0) {
            let l = lines[i].trim();
            if (l === '{') d++;
            else if (l === '}') d--;
            else if (d === 1) this.parseKV(l, obj.meta);
            i++;
        }
        return i;
    }

    static extractBlock(lines, startIdx) {
        let block = [];
        let d = 0;
        for (let i = startIdx; i < lines.length; i++) {
            let l = lines[i];
            let tl = l.trim();
            if (i === startIdx) { d = 1; continue; }
            if (tl === '{') d++;
            else if (tl === '}') d--;
            if (d === 0) break;
            block.push(l);
        }
        return block;
    }

    static processFormatting(text) {
        // 1. link{...}
        text = text.replace(/link\s*\{\s*display\s*:\s*"(.+?)"\s*,?\s*url\s*:\s*"(.+?)"\s*\}/g, '[$1]($2)');
        text = text.replace(/link\s*\{\s*url\s*:\s*"(.+?)"\s*,?\s*display\s*:\s*"(.+?)"\s*\}/g, '[$2]($1)');
        // 2. link:
        text = text.replace(/link:(.+?):(https?:\/\/\S+)/g, '[$1]($2)');
        // 3. strong{...}
        text = text.replace(/(?:strong|bold|b)\s*\{(.+?)\}/g, '**$1**');
        // 4. italic{...}
        text = text.replace(/(?:italic|it|i)\s*\{(.+?)\}/g, '*$1*');
        // 5. code{...}
        text = text.replace(/code\s*\{(.+?)\}/g, '`$1`');
        // 6. strong:
        text = text.replace(/(?:strong|bold|b):(.+)$/g, '**$1**');
        // 7. italic:
        text = text.replace(/(?:italic|it|i):(.+)$/g, '*$1*');
        // 8. code:
        text = text.replace(/code:(.+)$/g, '`$1`');
        return text;
    }

    static processContent(lines) {
        let out = "";
        let i = 0;
        while (i < lines.length) {
            let line = lines[i];
            let tl = line.trim();
            if (!tl || tl === '}') { i++; continue; }

            // Headers
            let m;
            if (m = tl.match(/^(h[1-6]):(.*)/)) {
                let lvl = m[1][1];
                let txt = m[2].trim().replace(/\s*\{$/, '');
                out += '#'.repeat(lvl) + ' ' + this.processFormatting(txt) + '\n\n';
                i++;
                continue;
            }
            // p{...}
            if (m = tl.match(/^p\s*\{(.+?)\}/)) {
                out += this.processFormatting(m[1]) + '\n\n';
                i++; continue;
            }
            // p block
            if (tl.match(/^p\s*\{/)) {
                let block = this.extractBlock(lines, i);
                block.forEach(l => { if(l.trim()) out += this.processFormatting(l.trim()) + '\n'; });
                out += '\n';
                i += block.length + 2; continue;
            }
            // p:
            if (tl.startsWith('p:')) {
                out += this.processFormatting(tl.substring(2).trim()) + '\n\n';
                i++; continue;
            }
            // lists
            if (m = tl.match(/^(ul|ol)\s*\{/)) {
                let type = m[1];
                let block = this.extractBlock(lines, i);
                let idx = 1;
                block.forEach(l => {
                    let itl = l.trim();
                    if (itl.startsWith('li:')) {
                        let bullet = type === 'ul' ? '-' : `${idx++}.`;
                        out += `${bullet} ${this.processFormatting(itl.substring(3).trim())}\n`;
                    }
                });
                out += '\n';
                i += block.length + 2; continue;
            }
            // checklist
            if (m = tl.match(/^(checklist|cl)\s*\{/)) {
                let block = this.extractBlock(lines, i);
                block.forEach(l => {
                    let itl = l.trim();
                    if (itl.match(/^(item\.checked|itm\.c):/)) out += `- [x] ${this.processFormatting(itl.split(':')[1].trim())}\n`;
                    else if (itl.match(/^(item|item\.unchecked|itm\.u):/)) out += `- [ ] ${this.processFormatting(itl.split(':')[1].trim())}\n`;
                });
                out += '\n';
                i += block.length + 2; continue;
            }
            // code block
            if (m = tl.match(/^code(?:\.(\w+))?\s*\{/)) {
                let lang = m[1] || '';
                let block = this.extractBlock(lines, i);
                let minIndent = Infinity;
                block.forEach(l => { if(l.trim()) minIndent = Math.min(minIndent, l.search(/\S/)); });
                out += '```' + lang + '\n';
                block.forEach(l => { out += (minIndent === Infinity ? l : l.substring(minIndent)) + '\n'; });
                out += '```\n\n';
                i += block.length + 2; continue;
            }
            // blockquote block
            if (m = tl.match(/^(blockquote|bq)\s*\{/)) {
                let block = this.extractBlock(lines, i);
                block.forEach(l => { if(l.trim()) out += `> ${this.processFormatting(l.trim())}\n`; });
                out += '\n';
                i += block.length + 2; continue;
            }
            // br:
            if (tl === 'br:') { out += '\n'; i++; continue; }
            // hr
            if (tl === 'hr') { out += '---\n\n'; i++; continue; }
            // img{...}
            if (m = tl.match(/^img\s*\{\s*(src|alt)\s*:\s*"(.+?)"\s*,?\s*(src|alt)\s*:\s*"(.+?)"\s*\}/)) {
                let data = {}; data[m[1]] = m[2]; data[m[3]] = m[4];
                out += `![${data.alt}](${data.src})\n\n`;
                i++; continue;
            }
            // img:
            if (tl.startsWith('img:')) {
                let parts = tl.substring(4).split(':');
                out += `![${parts[0]}](${parts.slice(1).join(':')})\n\n`;
                i++; continue;
            }
            // table
            if (tl.match(/^table\s*\{/)) {
                let block = this.extractBlock(lines, i);
                let rows = [];
                for(let j=0; j<block.length; j++) {
                    if(block[j].trim().match(/^tr\s*\{/)) {
                        let rowData = this.extractBlock(block, j);
                        let cells = rowData.map(c => this.processFormatting(c.trim().split(':')[1] || '').trim());
                        rows.push(cells);
                        j += rowData.length + 1;
                    }
                }
                if (rows.length > 0) {
                    out += `| ${rows[0].join(' | ')} |\n| ${rows[0].map(() => '---').join(' | ')} |\n`;
                    for(let r=1; r<rows.length; r++) out += `| ${rows[r].join(' | ')} |\n`;
                    out += '\n';
                }
                i += block.length + 2; continue;
            }
            // standalone
            if (tl.startsWith('blockquote:') || tl.startsWith('bq:')) {
                out += `> ${this.processFormatting(tl.substring(tl.indexOf(':')+1).trim())}\n\n`;
            } else if (tl.startsWith('li:')) {
                out += `- ${this.processFormatting(tl.substring(3).trim())}\n`;
            } else if (m = tl.match(/^code\.(\w+):(.*)/)) {
                out += '```' + m[1] + '\n' + m[2].trim() + '\n```\n\n';
            } else if (tl.startsWith('code:')) {
                out += '`' + tl.substring(5).trim() + '`' + '\n\n';
            }
            i++;
        }
        return out;
    }
}

if (process.argv.length >= 4) {
    Converter.convertCurlyPromptToSKILL(process.argv[2], process.argv[3]);
}

File: examples/skill_md_creator/spec/converter.py
import sys
import re
import math


class Converter:
    @staticmethod
    def convert_curly_prompt_to_skill(source_prompt_file, target_md_file):
        with open(source_prompt_file, 'r', encoding='utf-8') as f:
            source_prompt_content = f.read()
        lines = source_prompt_content.split('\n')
        output_lines = []
        frontmatter = {}
        content_lines = []
        in_skill = False
        in_content = False

        # Step 1: Parse the source file line by line
        i = 0
        while i < len(lines):
            line = lines[i].strip()

            if not in_skill and re.match(r'^skill\s*\{', line):
                in_skill = True
                i += 1
                continue

            if not in_skill:
                i += 1
                continue

            if not in_content:
                # Handle header { } block - contains frontmatter key-value pairs
                if re.match(r'^(?:header|head)\s*\{', line):
                    i += 1
                    while i < len(lines):
                        header_line = lines[i].strip()
                        if header_line == '}':
                            break
                        i = Converter._parse_frontmatter_line(header_line, frontmatter, lines, i)
                        i += 1
                    i += 1
                    continue

                # Parse frontmatter key-value pairs (flat structure, no header block)
                i = Converter._parse_frontmatter_line(line, frontmatter, lines, i)

                if re.match(r'^(?:content|body)\s*\{', line):
                    in_content = True
                    i += 1
                    continue
            else:
                # Collect all lines inside the content block into content_lines array
                # Store RAW (untrimmed) lines for code block indentation preservation
                brace_depth = 1
                j = i
                while j < len(lines):
                    inner_line_raw = lines[j]
                    inner_line = inner_line_raw.strip()
                    if inner_line == '}':
                        brace_depth -= 1
                        if brace_depth == 0:
                            break
                    else:
                        brace_depth += len(re.findall(r'\{', inner_line))
                        brace_depth -= len(re.findall(r'\}', inner_line))
                    content_lines.append(inner_line_raw)
                    j += 1
                i = j

            i += 1

        # Step 2: Build the frontmatter section
        output_lines.append('---')
        for key in frontmatter:
            if key == 'meta':
                output_lines.append('meta:')
                for attr_key in frontmatter['meta']:
                    output_lines.append('  ' + attr_key + ': ' + frontmatter['meta'][attr_key])
            else:
                output_lines.append(key + ': ' + frontmatter[key])
        output_lines.append('---')
        output_lines.append('')

        # Step 3: Process content_lines to generate markdown
        Converter._process_content_lines(content_lines, output_lines)

        # Step 4: Join output_lines with newline and add trailing newline
        target_md_content = '\n'.join(output_lines) + '\n'
        with open(target_md_file, 'w', encoding='utf-8') as f:
            f.write(target_md_content)

    @staticmethod
    def _parse_frontmatter_line(line, frontmatter, lines, i):
        if line.startswith('name:'):
            frontmatter['name'] = line[5:].strip().replace('"', '')
        elif line.startswith('description:'):
            frontmatter['description'] = line[12:].strip().replace('"', '')
        elif line.startswith('license:'):
            frontmatter['license'] = line[8:].strip().replace('"', '')
        elif line.startswith('version:'):
            frontmatter['version'] = line[8:].strip().replace('"', '')
        elif line.startswith('include:'):
            frontmatter['include'] = line[8:].strip().replace('"', '')
        elif re.match(r'^meta\s*\{', line):
            frontmatter['meta'] = {}
            j = i + 1
            while j < len(lines):
                meta_line = lines[j].strip()
                if meta_line == '}':
                    break
                colon_idx = meta_line.find(':')
                if colon_idx != -1:
                    key = meta_line[:colon_idx].strip()
                    value = meta_line[colon_idx + 1:].strip().replace('"', '')
                    frontmatter['meta'][key] = value
                j += 1
            return j
        return i

    @staticmethod
    def _process_formatting(text):
        # Process links: link{display:"..." url:"..."} (brace syntax with flexible spacing)
        def link_brace_replacer(match):
            inner = match.group(1)
            display = ''
            url = ''
            display_match = re.search(r'display:"([^"]+)"', inner)
            url_match = re.search(r'url:"([^"]+)"', inner)
            if display_match:
                display = display_match.group(1)
            if url_match:
                url = url_match.group(1)
            return '[' + display + '](' + url + ')'

        text = re.sub(r'link\s*\{([^}]+)\}', link_brace_replacer, text)

        # Process links: link:display text:url
        text = re.sub(r'link:(.+?):(https?://\S+)', r'[\1](\2)', text)

        # Process same-line open/close: strong{...}/bold{...}/b{...}, italic{...}/i{...}, code{...}
        text = re.sub(r'(?:strong|bold|b)\s*\{([^}]+)\}', r'**\1**', text)
        text = re.sub(r'(?:italic|it|i)\s*\{([^}]+)\}', r'*\1*', text)
        text = re.sub(r'code\s*\{([^}]+)\}', r'`\1`', text)

        # Process strong/bold/b: captures to end of line
        text = re.sub(r'(?:strong|bold|b):(.+)$', r'**\1**', text)

        # Process italic/i: captures to end of line
        text = re.sub(r'(?:italic|it|i):(.+)$', r'*\1*', text)

        # Process inline code: captures to end of line
        text = re.sub(r'code:\s*(.+)$', r'`\1`', text)

        return text

    @staticmethod
    def _process_content_lines(lines, output):
        i = 0
        while i < len(lines):
            line = lines[i].strip()

            if line == '' or line == '}':
                i += 1
                continue

            # Handle headers
            header_match = re.match(r'^(h[1-6]):(.*)', line)
            if header_match:
                level = int(header_match.group(1)[1])
                text = header_match.group(2).strip()
                if text.endswith('{'):
                    text = text[:-1].strip()
                text = Converter._process_formatting(text)
                output.append('#' * level + ' ' + text)
                output.append('')
                i += 1
                continue

            # Handle paragraphs (same-line form: p{...})
            p_inline_match = re.match(r'^p\s*\{(.+)\}$', line)
            if p_inline_match:
                text = Converter._process_formatting(p_inline_match.group(1))
                output.append(text)
                output.append('')
                i += 1
                continue

            # Handle paragraphs (block form: p { ... })
            if re.match(r'^p\s*\{', line):
                i += 1
                while i < len(lines):
                    inner = lines[i].strip()
                    if inner == '}':
                        break
                    if inner == 'br:':
                        output.append('')
                    elif inner != '':
                        text = Converter._process_formatting(inner)
                        output.append(text)
                    i += 1
                output.append('')
                i += 1
                continue

            # Handle paragraphs (inline form: p:TEXT)
            if line.startswith('p:'):
                text = line[2:].strip()
                text = Converter._process_formatting(text)
                output.append(text)
                output.append('')
                i += 1
                continue

            # Handle unordered lists
            if re.match(r'^ul\s*\{', line):
                i += 1
                while i < len(lines):
                    inner = lines[i].strip()
                    if inner == '}':
                        break
                    if inner.startswith('li:'):
                        text = inner[3:].strip()
                        text = Converter._process_formatting(text)
                        output.append('- ' + text)
                    i += 1
                output.append('')
                i += 1
                continue

            # Handle ordered lists
            if re.match(r'^ol\s*\{', line):
                i += 1
                num = 1
                while i < len(lines):
                    inner = lines[i].strip()
                    if inner == '}':
                        break
                    if inner.startswith('li:'):
                        text = inner[3:].strip()
                        text = Converter._process_formatting(text)
                        output.append(str(num) + '. ' + text)
                        num += 1
                    i += 1
                output.append('')
                i += 1
                continue

            # Handle code blocks (with optional language: code { } or code.<lang> { })
            code_block_match = re.match(r'^code(?:\.(\w+))?\s*\{', line)
            if code_block_match:
                lang = code_block_match.group(1) or ''
                i += 1
                code_lines = []
                depth = 1
                while i < len(lines):
                    inner = lines[i].strip()
                    if inner == '}':
                        depth -= 1
                        if depth == 0:
                            break
                        code_lines.append(lines[i])
                    else:
                        if inner != '':
                            depth += len(re.findall(r'\{', inner))
                            depth -= len(re.findall(r'\}', inner))
                        code_lines.append(lines[i])
                    i += 1
                # Strip common indentation
                min_indent = math.inf
                for cl in code_lines:
                    if cl.strip() == '':
                        continue
                    indent = len(re.match(r'^(\s*)', cl).group(1))
                    if indent < min_indent:
                        min_indent = indent
                if min_indent == math.inf:
                    min_indent = 0
                else:
                    min_indent = int(min_indent)
                output.append('```' + lang)
                for cl in code_lines:
                    output.append(cl[min_indent:])
                output.append('```')
                output.append('')
                i += 1
                continue

            # Handle multi-line blockquotes (blockquote { } or bq { })
            if re.match(r'^(?:blockquote|bq)\s*\{', line):
                i += 1
                while i < len(lines):
                    inner = lines[i].strip()
                    if inner == '}':
                        break
                    if inner != '':
                        text = Converter._process_formatting(inner)
                        output.append('> ' + text)
                    i += 1
                output.append('')
                i += 1
                continue

            # Handle blockquotes (blockquote: or bq:)
            if line.startswith('blockquote:') or line.startswith('bq:'):
                prefix = 'blockquote:' if line.startswith('blockquote:') else 'bq:'
                text = line[len(prefix):].strip()
                text = Converter._process_formatting(text)
                output.append('> ' + text)
                output.append('')
                i += 1
                continue

            # Handle line breaks
            if line == 'br:':
                output.append('')
                i += 1
                continue

            # Handle horizontal rules
            if line == 'hr':
                output.append('---')
                output.append('')
                i += 1
                continue

            # Handle images (brace syntax)
            img_brace_match = re.match(r'^img\s*\{(.+)\}$', line)
            if img_brace_match:
                inner = img_brace_match.group(1)
                src = ''
                alt = ''
                src_match = re.search(r'src:"([^"]+)"', inner)
                alt_match = re.search(r'alt:"([^"]+)"', inner)
                if src_match:
                    src = src_match.group(1)
                if alt_match:
                    alt = alt_match.group(1)
                output.append('![' + alt + '](' + src + ')')
                output.append('')
                i += 1
                continue
            if line.startswith('img:'):
                rest = line[4:]
                colon_idx = rest.index(':')
                alt = rest[:colon_idx]
                path = rest[colon_idx + 1:]
                output.append('![' + alt + '](' + path + ')')
                output.append('')
                i += 1
                continue

            # Handle tables
            if re.match(r'^table\s*\{', line):
                rows = []
                i += 1
                while i < len(lines):
                    inner = lines[i].strip()
                    if inner == '}':
                        break
                    if re.match(r'^tr\s*\{', inner):
                        cells = []
                        is_header_row = False
                        i += 1
                        while i < len(lines):
                            cell_line = lines[i].strip()
                            if cell_line == '}':
                                break
                            if cell_line.startswith('th:'):
                                is_header_row = True
                                text = cell_line[3:].strip()
                                text = Converter._process_formatting(text)
                                cells.append(text)
                            elif cell_line.startswith('td:'):
                                text = cell_line[3:].strip()
                                text = Converter._process_formatting(text)
                                cells.append(text)
                            i += 1
                        rows.append({'cells': cells, 'is_header': is_header_row})
                    i += 1
                if rows:
                    first_row = rows[0]
                    col_count = len(first_row['cells'])
                    # Emit header row
                    output.append('| ' + ' | '.join(first_row['cells']) + ' |')
                    # Emit separator
                    output.append('| ' + ' | '.join(['---'] * col_count) + ' |')
                    # Emit data rows (skip first row since it's used as header)
                    for r in range(1, len(rows)):
                        output.append('| ' + ' | '.join(rows[r]['cells']) + ' |')
                    output.append('')
                i += 1
                continue

            # Handle checklists (checklist or cl)
            if re.match(r'^(?:checklist|cl)\s*\{', line):
                i += 1
                while i < len(lines):
                    inner = lines[i].strip()
                    if inner == '}':
                        break
                    # item.checked / itm.c
                    checked_match = re.match(r'^(?:item\.checked|itm\.c):', inner)
                    if checked_match:
                        text = inner[checked_match.end():].strip()
                        text = Converter._process_formatting(text)
                        output.append('- [x] ' + text)
                    else:
                        # item.unchecked / itm.u / item (plain)
                        unchecked_match = re.match(r'^(?:item\.unchecked|itm\.u|item):', inner)
                        if unchecked_match:
                            text = inner[unchecked_match.end():].strip()
                            text = Converter._process_formatting(text)
                            output.append('- [ ] ' + text)
                    i += 1
                output.append('')
                i += 1
                continue

            # Handle standalone list items
            if line.startswith('li:'):
                text = line[3:].strip()
                text = Converter._process_formatting(text)
                output.append('- ' + text)
                i += 1
                continue

            # Handle standalone code.<lang>:TEXT (single-line language code block)
            code_lang_inline_match = re.match(r'^code\.(\w+):(.+)$', line)
            if code_lang_inline_match:
                lang = code_lang_inline_match.group(1)
                text = code_lang_inline_match.group(2).strip()
                output.append('```' + lang)
                output.append(text)
                output.append('```')
                output.append('')
                i += 1
                continue

            # Handle standalone code: lines (inline code as paragraph)
            if line.startswith('code:'):
                text = line[5:].strip()
                output.append('`' + text + '`')
                output.append('')
                i += 1
                continue

            i += 1


# CLI entry point
if __name__ == '__main__':
    if len(sys.argv) >= 3:
        Converter.convert_curly_prompt_to_skill(sys.argv[1], sys.argv[2])
